lib/src/functions/math.dart

```diff
@@ -13,34 +13,58 @@ import '../module/built_in.dart';
 import '../util/number.dart';
 import '../value.dart';
 
-/// A random number generator.
-final _random = math.Random();
-
 /// The global definitions of Sass math functions.
 final global = UnmodifiableListView([
-  _round, _ceil, _floor, _abs, _max, _min, _randomFunction, _unit, //
-  _percentage,
+  _abs, _ceil, _floor, _max, _min, _percentage, _randomFunction, _round,
+  _unit, //
+  _compatible.withName("comparable"),
   _isUnitless.withName("unitless"),
-  _compatible.withName("comparable")
 ]);
 
 /// The Sass math module.
 final module = BuiltInModule("math", functions: [
-  _round, _ceil, _floor, _abs, _max, _min, _randomFunction, _unit,
-  _isUnitless, //
-  _percentage, _compatible
+  _abs, _ceil, _clamp, _compatible, _floor, _hypot, _isUnitless, _max, _min, //
+  _percentage, _randomFunction, _round, _unit,
 ]);
 
-final _percentage = BuiltInCallable("percentage", r"$number", (arguments) {
-  var number = arguments[0].assertNumber("number");
-  number.assertNoUnits("number");
-  return SassNumber(number.value * 100, '%');
-});
+/// Returns a [Callable] named [name] that transforms a number's value
+/// using [transform] and preserves its units.
+BuiltInCallable _numberFunction(String name, num transform(num value)) {
+  return BuiltInCallable(name, r"$number", (arguments) {
+    var number = arguments[0].assertNumber("number");
+    return SassNumber.withUnits(transform(number.value),
+        numeratorUnits: number.numeratorUnits,
+        denominatorUnits: number.denominatorUnits);
+  });
+}
+
+///
+/// Bounding functions
+///
 
-final _round = _numberFunction("round", fuzzyRound);
 final _ceil = _numberFunction("ceil", (value) => value.ceil());
+
+final _clamp = BuiltInCallable("clamp", r"$min, $number, $max", (arguments) {
+  var min = arguments[0].assertNumber("min");
+  var number = arguments[1].assertNumber("number");
+  var max = arguments[2].assertNumber("max");
+
+  if (min.hasUnits == number.hasUnits && number.hasUnits == max.hasUnits) {
+    if (min.greaterThanOrEquals(max).isTruthy) return min;
+    if (min.greaterThanOrEquals(number).isTruthy) return min;
+    if (number.greaterThanOrEquals(max).isTruthy) return max;
+    return number;
+  }
+
+  var arg1 = "$min";
+  var arg2 = "${min.hasUnits != number.hasUnits ? number : max}";
+  var unit1 = "${min.hasUnits ? "has units" : "is unitless"}";
+  var unit2 = "${min.hasUnits ? "is unitless" : "has units"}";
+  throw SassScriptException(
+      "$arg1 $unit1 but $arg2 $unit2. Arguments must all have units or all be unitless.");
```

Comment:
I'd also include the names of the arguments in question here—it may not be obvious from context.

Also, this is a long line (the formatter won't break strings for you).

Same below.

Reply:
Done.